src/frontend/src/optimizer/plan_node/logical_filter.rs

```diff
@@ -185,11 +190,120 @@ impl ToBatch for LogicalFilter {
-        let new_logical = self.clone_with_input(new_input);
-        Ok(StreamFilter::new(new_logical).into())
+
+        let predicate = self.predicate();
+        let has_now = predicate
+            .conjunctions
+            .iter()
+            .any(|cond| cond.count_nows() > 0);
+        if has_now {
+            let mut conjunctions = predicate.conjunctions.clone();
+            // Check if the now expr is valid
+            for conjunction in &conjunctions {
+                if conjunction.count_nows() > 0 {
+                    let comparator_expr = try_match_expand!(conjunction, ExprImpl::FunctionCall)?;
+                    if convert_comparator_to_priority(comparator_expr.get_expr_type()) < 0 {
+                        return Err(ExprError::InvalidParam {
+                            name: "now",
+                            reason: String::from("now expression must be placed in a comparison"),
+                        }
+                        .into());
+                    }
+                    try_match_expand!(&comparator_expr.inputs()[0], ExprImpl::InputRef)?;
+                    let now_expr =
+                        try_match_expand!(&comparator_expr.inputs()[1], ExprImpl::FunctionCall)?;
+                    match now_expr.get_expr_type() {
+                        Type::Now => {
+                            // Do nothing.
+                        }
+                        Type::Add | Type::Subtract => {
+                            if try_match_expand!(&now_expr.inputs()[0], ExprImpl::FunctionCall)?
+                                .get_expr_type()
+                                != Type::Now
+                            {
+                                return Err(ExprError::InvalidParam {
+                                    name: "now",
+                                    reason: String::from("now() must appear in the left side of the now delta expression directly"),
+                                }
+                                .into());
+                            }
+                        }
+                        _ => {
+                            return Err(ExprError::InvalidParam {
+                                name: "now",
+                                reason: String::from("now delta expression must be a trivial add/subtract expression"),
+                            }
+                            .into());
+                        }
+                    }
+                }
+            }
+
+            let mut now_conds = conjunctions
+                .drain_filter(|cond| cond.count_nows() > 0)
+                .map(|cond| {
+                    must_match!(cond, ExprImpl::FunctionCall(function_call) => {
+                        (convert_comparator_to_priority(function_call.get_expr_type()), function_call)
+                    })
+                })
+                .collect_vec();
+            now_conds.sort_by_key(|(comparator_priority, _)| *comparator_priority);
```

Comment:
Let's add some comments. 

// Apply equality join first, then conditions of the form `T > now() - Y` (the timestamp needs to be greater than a watermark), then conditions similar to `X < now() - Y`

Reply:
> Let's add some comments.
> 
> // Apply filters by selectivity and then applicabiliy of watermark - equality condition first, then conditions of the form `T > now() - Y` (the timestamp needs to be greater than a watermark), then conditions similar to `T < now() - Y`

fixed